Fix GH Actions job to run on Ubuntu 24.04

enkiusz opened the pull request:
Github is scheduled to finish upgrading the 'ubuntu-latest' workers to Ubuntu 24.04 on 2025-01-17 (https://github.com/actions/runner-images/issues/10636).

Current docbuild job fails on a ubuntu-24.04 runner as can be evidenced by a failed run on my fork (https://github.com/enkiusz/connectedhomeip/actions/runs/12765165078/job/35578813277).
The reason for this is that built-in pip is no longer allowed to upgrade itself. I altered this job to use a venv instead.

Additionally, I bumped the Python version.

#### Testing

The new job has ran successfully on ubuntu-24.04 (https://github.com/enkiusz/connectedhomeip/actions/runs/12767259766/job/35585278460) and ubuntu 22.04 (https://github.com/enkiusz/connectedhomeip/actions/runs/12767352931/job/35585563551).

enkiusz committed:
docbuild: Run sphinx in venv, update python version

Pip cannot upgrade pip on Ubuntu 24.04. Instead create and use a venv to run
the sphinx job. Also bump Python version to 3.12.

Signed-off-by: Maciej Grela <m.grela@samsung.com>

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/project-chip/connectedhomeip?pullRequest=37063) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/project-chip/connectedhomeip?pullRequest=37063) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/project-chip/connectedhomeip?pullRequest=37063) it.</sub>

arkq commented on .github/workflows/docbuild.yaml:
```diff
@@ -29,17 +29,19 @@ jobs:
-          sudo pip3 install -U pip
-          pip3 install -r docs/requirements.txt
+          python3 -m venv venv
+          venv/bin/pip3 install -U pip
```
NIT: How about skipping pip update? The pip version shipped with python should be good enough to install everything. If the pip is too old for something, it's rather an indication that the Python version is outdated...

enkiusz replied:
I agree, let's skip this.

enkiusz committed:
Merge branch 'project-chip:master' into docbuild-ubuntu-24.04

enkiusz committed:
Post-review updates:

- do not update pip
- use a venv also in cancel_workflows_for_pr.yaml and protocol_compatibility.yaml
- bump python version where applicable

Signed-off-by: Maciej Grela <m.grela@samsung.com>
